### PR TITLE
networkd: ndisc - do not stop ndisc client incase of conf error.

### LINKED ISSUE #5625: systemd-networkd won't renew DHCPv4 address after NDisc error
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

systemd 232

### Used distribution

Arch Linux ARM

### In case of bug report: Expected behaviour you didn't see

When DHCPv4 lease expires, the address should be renewed from the DHCP server.

### In case of bug report: Unexpected behaviour you saw

When DHCPv4 lease expired, the address was not renewed and the machine lost IPv4 connectivity.  IPv6 connectivity remained.

### In case of bug report: Steps to reproduce the problem

I am not sure how to reproduce the problem as I do not know what is causing the error message.

I see this in my logs when starting `systemd-networkd`:

```
systemd-networkd[245]: Enumeration completed
systemd-networkd[245]: eth0: IPv6 enabled for interface: Success
systemd-networkd[245]: eth0: Gained carrier
systemd-networkd[245]: eth0: Lost carrier
systemd-networkd[245]: eth0: Gained IPv6LL
systemd-networkd[245]: eth0: Gained carrier
systemd-networkd[245]: eth0: DHCPv4 address 192.168.0.25/24 via 192.168.0.10
systemd-networkd[245]: eth0: Could not set NDisc route or address: Invalid argument
systemd-networkd[245]: eth0: Failed
```
Then 24 hours later (the length of the IPv4 lease) the IPv4 address is removed again with nothing written to the logs.  This has happened on three different machines.

`systemctl status systemd-networkd` shows it as active and running.

Restarting `systemd-networkd` restores the IPv4 address, and displays the following log messages:
```
systemd-networkd[750]: eth0: Gained IPv6LL
systemd-networkd[750]: Enumeration completed
systemd-networkd[750]: eth0: DHCPv4 address 192.168.0.25/24 via 192.168.0.10
systemd-networkd[750]: eth0: Could not set NDisc route or address: Invalid argument
systemd-networkd[750]: eth0: Failed
```
I presume in another 24 hours the IPv4 address will drop out again.

I am not sure what it causing the `NDisc` error message.  I have IPv6 enabled on this network so it may be related to that, however any IPv6 errors should not prevent IPv4 addresses from being renewed.

Please advise if there are further troubleshooting steps I can take to help narrow down the problem.

Here is the content of my `/etc/systemd/network/default.network` file, the only one I have in that directory:
```
[Match]
Name=eth*

[Network]
DHCP=v4

Address=fdaa::ba27:ebff:fe48:1c6b

[DHCPv4]
UseHostname=false
```

### PR DESCRIPTION
**Pull request by ssahani**
Now in ndisc_netlink_handler if route or address fails we stop the clients.
link_enter_failed->link_stop_clients that is dhcp, ndisc etc.

The clients should be keep on running .

Fixes #5625

### CODE REVIEW

File: src/network/networkd-ndisc.c [hunk 1]
```diff
@@ -39,10 +39,8 @@ static int ndisc_netlink_handler(sd_netlink *rtnl, sd_netlink_message *m, void *
         link->ndisc_messages--;
 
         r = sd_netlink_message_get_errno(m);
-        if (r < 0 && r != -EEXIST) {
+        if (r < 0 && r != -EEXIST)
                 log_link_error_errno(link, r, "Could not set NDisc route or address: %m");
```
Comment (poettering) [inline src/network/networkd-ndisc.c:43]: hmm, but why does this fail in the first place?

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -755,7 +755,7 @@ void link_check_ready(Link *link) {
              !link->dhcp4_configured && !link->dhcp6_configured))
                 return;
 
-        if (link_ipv6_accept_ra_enabled(link) && !link->ndisc_configured)
+        if (link_ipv6_accept_ra_enabled(link) && !link->ndisc_configured && !link->network->bridge)
```
Comment (poettering) [inline src/network/networkd-link.c]: can you elaborate why we only check this for IPv6 RA here? wouldn't it make sense to skip the whole section (including the dhcp/ipv4 stuff) if link->network->bridge is set?

Comment (ssahani) [reply src/network/networkd-link.c]: yes since we disable both .

File: src/network/networkd-ndisc.c (hunk 1 above)
Comment (ssahani) [reply src/network/networkd-ndisc.c:43]: It seems a multiple reasons here prefix change one of them .

**Comment by ssahani** [comment]
Updated.

File: src/network/networkd-ndisc.c (hunk 1 above)
Comment (ssahani) [reply src/network/networkd-ndisc.c:43]: Also I think we should not bring down the dhcp client too when some error happen. The admin could do a wrong conf and in that case whole network will go down where networkd is running. Then admin has to go each of the machines and start manually. I think we should log and keep on running the network clients.

**Comment by mikken** [comment]
I tested this patch and DHCPv4 client no longer breaks.

**Comment by ssahani** [comment]
thanks but what if there is issue with dhcp conf can't set set correctly and there we too stopping the client . I think that needs to be fixed too.

**Pull request merged**